### ISSUE TITLE
Using cache cause random behavior during generation

### ISSUE THREAD

**Issue by Larryvrh** [turboderp/exllama#73]
I'm currently testing the different generation behavior between exllama and autogptq, and I found that using cache with exllama will generate different results for same prompt even when I'm using greedy decoding.

```python
def exllama_greedy_gen_wo_cache(prompt, max_length):
    seq = tokenizer.encode(prompt) # Huggingface tokenizer
    for _ in range(max_length):
        temp_cache = ExLlamaCache(model_exllama)
        logits = model_exllama.forward(torch.tensor([seq], dtype=torch.long), temp_cache)[0][0]
        seq.append(torch.argmax(logits).item())
    return seq

for i in range(10):
    print(tokenizer.decode(exllama_greedy_gen_wo_cache("Hello,", 20)))
```
For generate without cache, it's really slow but I get consistant outputs
![图片](https://github.com/turboderp/exllama/assets/15991834/cc83da0b-c736-4606-adf3-540d3edb6a32)

But when I enable cache to get a much faster generation, I start seeing inconsistency between generations

```python
def exllama_greedy_gen_wi_cache(prompt, max_length):
    seq = tokenizer.encode(prompt) # Huggingface tokenizer
    gen_cache = ExLlamaCache(model_exllama)
    model_exllama.forward(torch.tensor([seq[:-1]], dtype=torch.long), gen_cache, preprocess_only = True)
    for _ in range(max_length):
        logits = model_exllama.forward(torch.tensor([seq[-1:]], dtype=torch.long), gen_cache)[0][0]
        seq.append(torch.argmax(logits).item())
    return seq

for i in range(10):
    print(tokenizer.decode(exllama_greedy_gen_wi_cache("Hello,", 20)))
```
![图片](https://github.com/turboderp/exllama/assets/15991834/75aa99f4-daa0-49db-977e-4840ab6ce0bf)

I wonder is this a bug within the cache implementation or it is I'm using cache in a wrong way.

**Comment by turboderp** [comment]
There doesn't seem to be anything wrong with the generator loops. It should be a valid way of using it.

I tried to replicate it using a couple of different models, using either ExLlamaTokenizer or HF tokenizer, and I'm not getting the same behavoir. Here's the full code including setup:

```Python
from model import ExLlama, ExLlamaCache, ExLlamaConfig
from tokenizer import ExLlamaTokenizer
import torch
from transformers import LlamaTokenizer
import os, glob

torch.set_grad_enabled(False)
torch.cuda._lazy_init()

model_directory =  "/mnt/str/models/_test_models/Neko-Institute-of-Science_LLaMA-13B-4bit-128g"

tokenizer_path = os.path.join(model_directory, "tokenizer.model")
model_config_path = os.path.join(model_directory, "config.json")
st_pattern = os.path.join(model_directory, "*.safetensors")
model_path = glob.glob(st_pattern)[0]

config = ExLlamaConfig(model_config_path)
config.model_path = model_path
model_exllama = ExLlama(config)

tokenizer = LlamaTokenizer.from_pretrained(model_directory)  # Preferred method from HF
# tokenizer = LlamaTokenizer.from_pretrained(tokenizer_path)  # Works, but deprecated
hf_tokenizer = True

# tokenizer = ExLlamaTokenizer(tokenizer_path)
# hf_tokenizer = False

def exllama_greedy_gen_wo_cache(prompt, max_length):
    if hf_tokenizer: seq = tokenizer.encode(prompt)
    else: seq = [1] + tokenizer.encode(prompt).squeeze(0).tolist()
    for _ in range(max_length):
        temp_cache = ExLlamaCache(model_exllama)
        logits = model_exllama.forward(torch.tensor([seq], dtype=torch.long), temp_cache)[0][0]
        seq.append(torch.argmax(logits).item())
    return seq if hf_tokenizer else torch.tensor(seq, dtype=torch.long)

for i in range(10):
    print(tokenizer.decode(exllama_greedy_gen_wo_cache("Hello,", 20)))

print("---")

def exllama_greedy_gen_wi_cache(prompt, max_length):
    if hf_tokenizer: seq = tokenizer.encode(prompt)
    else: seq = [1] + tokenizer.encode(prompt).squeeze(0).tolist()
    gen_cache = ExLlamaCache(model_exllama)
    model_exllama.forward(torch.tensor([seq[:-1]], dtype=torch.long), gen_cache, preprocess_only=True)
    for _ in range(max_length):
        logits = model_exllama.forward(torch.tensor([seq[-1:]], dtype=torch.long), gen_cache)[0][0]
        seq.append(torch.argmax(logits).item())
    return seq if hf_tokenizer else torch.tensor(seq, dtype=torch.long)

for i in range(10):
    print(tokenizer.decode(exllama_greedy_gen_wi_cache("Hello,", 20)))
```

Produces 10x this, for both loops:

`<s>Hello, I'm interested in "The Art of War" Artwork. Could you please confirm its avail`

Different results for different models, but always consistent.

I notice that you're seeing a difference in the _first_ part of the string. `<s>Hello,` vs. `<s> Hello,`. This part isn't sampled from the logits but comes straight out of the tokenizer. In my case, the prompt is consistently tokenized as `[1, 15043, 29892]`, whether I use the HF tokenizer or not (explicitly adding the `[1]` token). But I'm wondering if for some reason you're maybe getting a different BOS token, or the "Hello" token alternates between two different versions for some reason. What do you get if you add `print(seq)` before the second line in each of your gen functions?

**Comment by Larryvrh** [comment]
![图片](https://github.com/turboderp/exllama/assets/15991834/518175e3-a669-41ac-8c1f-e0bf35eb60c6)
The seq seems indeed the same, i don't know where does the extra space come from. The model I'm using for the test is "TheBloke/tulu-7B-GPTQ."

**Comment by Larryvrh** [comment]
Update: I tried the "Neko-Institute-of-Science/LLaMA-13B-4bit-128g" on my side, and it seems like this model is working properly and give consistent outputs. The model weight seems have something to do with the inconsistency.

**Comment by turboderp** [comment]
That's very strange. To sum up:

The HF tokenizer encodes the sequence `Hello,` to `[1, 15043, 29892]`, which then decodes to either `<s>Hello,` or `<s> Hello,`, apparently at random. But in those cases where it decodes to the second version, the model treats the _same_ three tokens differently for some reason.

I have to think there's something else that's different about the initial three tokens. Maybe the HF tokenizer sometimes outputs a different datatype? Are they floats in one version and ints in another, or something? 

What happens if you produce a tensor directly from the tokenizer with `return_tensors="pt"`?

What if you supply `add_special_tokens=False`?

I tried replicating the behavior with TheBloke/tulu-7B-GPTQ, but still nothing. Maybe I'm using the HF tokenizer differently? Cause there's some sort of interaction between the model and the tokenizer for sure, since it's not _just_ the model output that changes but also how the tokenizer decodes the same (?) token IDs afterwards. Could you write a more complete, minimal example that shows the same behavior?

**Comment by Larryvrh** [comment]
This is the full script I use
```python
import sys
import torch
from transformers import LlamaTokenizer

sys.path.insert(0, "exllama")
from exllama.model import ExLlama, ExLlamaConfig, ExLlamaCache

model_path = '/home/LinuxExtra/text-generation-webui/models/TheBloke_tulu-7B-GPTQ'
weight_name = 'gptq_model-4bit-128g.safetensors'


def get_exllama():
    config = ExLlamaConfig(model_path + '/config.json')
    config.model_path = str(model_path + '/' + weight_name)
    model = ExLlama(config)
    return model


tokenizer = LlamaTokenizer.from_pretrained(model_path, use_fast=False)
model_exllama = get_exllama()


def exllama_greedy_gen_wi_cache(prompt, max_length):
    gen_logits = []
    seq = tokenizer.encode(prompt)
    print('Start Seq:', seq)
    gen_cache = ExLlamaCache(model_exllama)
    model_exllama.forward(torch.tensor([seq[:-1]], dtype=torch.long), gen_cache, preprocess_only=True)
    for _ in range(max_length):
        logits = model_exllama.forward(torch.tensor([seq[-1:]], dtype=torch.long), gen_cache)[0][0]
        seq.append(torch.argmax(logits).item())
        gen_logits.append(logits[seq[-1]].item())
    print('Result Text:', [tokenizer.decode(seq)])
    print('Result Seq:', [(t, l) for t, l in zip(seq[-max_length:], gen_logits)])
    print('-' * 50)
    return seq


for i in range(10):
    exllama_greedy_gen_wi_cache("Hello,", 20)

```
![图片](https://github.com/turboderp/exllama/assets/15991834/daf463e9-e834-4102-a0f0-f5d3ca0f5339)
The result seems like simply there are two logits get really close values and the generation start altering between them?

**Comment by fxmarty** [comment]
@turboderp Is there any source of randomicity in the exllama gemm/gemv kernels? For some reason in my case, using exllama kernels between different runs with the same prompt + greedy search the output logits are slightly different, though the argmax is the same.

I don't hit this issue with the triton kernel from gptq-for-llama. I'll try to come up with a reproducible example.
